Swift: New query: Incomplete regular expression for hostnames

geoffw0 opened the pull request:
Adds a new Swift query "Incomplete regular expression for hostnames".  This is a port of `js/incomplete-hostname-regexp`; Ruby, Go and Python have similar queries as well.
- we use the shared logic in `shared/regex/codeql/regex/HostnameRegexp` for the query, with a thin layer in `swift/ql/lib/codeql/swift/security/regex/HostnameRegex.qll` connecting that to the Swift libraries.
- we use the existing Swift regex library `swift/ql/lib/codeql/swift/regex/Regex.qll`.  `ParsedStringRegex` is changed to extend a new class `RegexPatternSource` and ultimately `DataFlow::Node` rather than `Expr`, as that is the configuration of classes the shared library expects to find.
- tests and qhelp are similar to other languages though there has been some adapting, especially in the qhelp examples.

Before merging:
- [x] DCA run
- [ ] team review
- [x] docs review

MathiasVP commented on swift/ql/lib/codeql/swift/regex/Regex.qll:
```diff
@@ -8,23 +8,56 @@ private import codeql.swift.dataflow.DataFlow
-  DataFlow::Node getUse() { result = use }
+  override DataFlow::Node getAParse() { result = use }
+
+  override RegExpTerm getRegExpTerm() { result.getRegExp() = this.asExpr() }
```
```suggestion
  override RegExpTerm getRegExpTerm() { result.getRegExp() = expr }
```

geoffw0 replied:
Thanks - that's why I'm getting warnings from QL-for-QL as well.

MathiasVP reviewed (commented):
I've only reviewed the QL code so far.

geoffw0 commented on swift/ql/lib/codeql/swift/security/regex/HostnameRegex.qll:
```diff
@@ -0,0 +1,21 @@
+/**
+ * An implementation of the signature that allows the Hostname analysis to run.
+ */
+module Impl implements Shared::HostnameRegexpSig<TreeImpl> {
```
Yep!  (done)

geoffw0 commented:
DCA LGTM.  It shows a slight (~3%) overall analysis slowdown, which may be wobble or the real effect of adding a new query.  I will check the most slowed down project (`signalapp__Signal-iOS`) locally for any signs of an actual problem.

mchammer01 commented on swift/ql/src/queries/Security/CWE-020/IncompleteHostnameRegex.ql:
```diff
@@ -0,0 +1,16 @@
+/**
+ * @name Incomplete regular expression for hostnames
+ * @description Matching a URL or hostname against a regular expression that contains an unescaped dot as part of the hostname might match more hostnames than expected.
```
The description looks fine, but perhaps we could replace the second occurrence of "to match" so that there's less repetition. I've used "to return" but perhaps you can find a better synonym. Also used the present for "may".
```suggestion
 * @description Matching a URL or hostname against a regular expression that contains an unescaped dot as part of the hostname may return more hostnames than expected.
```

geoffw0 replied:
:+1: "may", I've made that change.

I'm not quite happy with the word "return" here since we don't know whether the user code we've identified is really "returning" anything (or whether it's doing something else with the match).  Sadly I can't find a good alternative word so I've left the repetition of "match" for now.

mchammer01 replied:
@geoffw0 - that's fine, let's lave it as-is. Thanks for addressing my other comments 👍🏻

mchammer01 commented on swift/ql/src/change-notes/2023-08-23-incomplete-hostname-regex.md:
```diff
@@ -0,0 +1,5 @@
+category: newQuery
+---
+
+* Added new query "Incomplete regular expression for hostnames" (`swift/incomplete-hostname-regexp`). This query finds regular expressions matching a URL or hostname that might match more hostnames than expected.
```
Let's use the present here.
```suggestion
* Added new query "Incomplete regular expression for hostnames" (`swift/incomplete-hostname-regexp`). This query finds regular expressions matching a URL or hostname that may match more hostnames than expected.
```

mchammer01 commented on swift/ql/src/queries/Security/CWE-020/IncompleteHostnameRegex.qhelp:
```diff
@@ -0,0 +1,74 @@
+
+	<references>
+		<li>OWASP: <a href="https://www.owasp.org/index.php/Server_Side_Request_Forgery">Server Side Request Forgery</a></li>
+		<li>OWASP: <a href="https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html">Unvalidated Redirects and Forwards Cheat Sheet</a></li>
```
Adding a full stop.
```suggestion
		<li>OWASP: <a href="https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html">Unvalidated Redirects and Forwards Cheat Sheet</a>.</li>
```

mchammer01 commented on swift/ql/src/queries/Security/CWE-020/IncompleteHostnameRegex.qhelp:
```diff
@@ -0,0 +1,74 @@
+	</example>
+
+	<references>
+		<li>OWASP: <a href="https://www.owasp.org/index.php/Server_Side_Request_Forgery">Server Side Request Forgery</a></li>
```
Add full stop.
```suggestion
		<li>OWASP: <a href="https://www.owasp.org/index.php/Server_Side_Request_Forgery">Server Side Request Forgery</a>.</li>
```

mchammer01 commented on swift/ql/src/queries/Security/CWE-020/IncompleteHostnameRegex.qhelp:
```diff
@@ -0,0 +1,74 @@
+
+		</p>
+
+		<sample src="IncompleteHostnameRegexBad.swift"/>
```
Wondering if we should add a BAD comment here, on the `let regex` line and a GOOD comment in the example where we're escaping the `.`?

geoffw0 replied:
Good idea.  It's a long enough example the reader might want something to guide their eye to the important line.

mchammer01 commented on swift/ql/src/queries/Security/CWE-020/IncompleteHostnameRegex.qhelp:
```diff
@@ -0,0 +1,74 @@
+
+		<p>
+
+			The check is however easy to bypass because the unescaped
```
However is in the middle of a clause, therefore it requires commas.
```suggestion
			The check is, however, easy to bypass because the unescaped
```

mchammer01 reviewed (dismissed):
@geoffw0 - this LGTM ✨ 
Added a few editorial suggestions (feel free to ignore the ones you don't agree with).

geoffw0 commented:
@mchammer01 thanks for the speedy review!  I've accepted or addressed all your comments, with one slight exception (see above).  Please say if you have any more suggestions there or elsewhere.

geoffw0 commented:
Locally, on the `Signal-iOS` snapshot, the query took 7 seconds (cache warmed up with another regex query).  Nothing stands out in the log.  I don't think there's a performance issue.

geoffw0 commented:
I think we just need someone from the team to have a quick read of the `.qhelp` as well, and give the :+1: overall.

mchammer01 reviewed (approved):
LGTM from an editorial perspective ✨

MathiasVP reviewed (approved):
LGTM!
